### DISCUSSION TITLE
adding provider to web3js

### DISCUSSION THREAD

**Discussion by mostafa-rio** [WalletConnect/walletconnect-monorepo#2920] [Developer Support] [answered]
I'm using web3js for sending transactions and web3modal for connecting to wallets in an angular app.
In the process of upgrading to web3modal v2, I need to pass the provider ( after user is connected by web3modal ) to web3.js.
( also checkout this [discussion](https://github.com/orgs/WalletConnect/discussions/1993) )

here is the code: 

```
import { EthereumProvider } from '@walletconnect/ethereum-provider;

const provider = await EthereumProvider.init({
      projected: 'projectid'
      chains: [1, 137, 56], // REQUIRED chain ids
      showQrModal: true,
      rpcMap: {
        1: "https://mainnet.infura.io/v3/"+'3212321321',
        56: "https://bsc-dataseed.binance.org",
        137: "https://rpc-mainnet.maticvigil.com",
      }
    })
    
      await this.provider.enable()
        this.web3js = new Web3(this.provider);

```

after this operation the error is:
```
Error: Unsupported chains.
    at index.es.js:1:3345
    at ZoneDelegate.invoke (zone.js:372:26)

ERROR Error: Uncaught (in promise): Error: Provider not set or invalid
Error: Provider not set or invalid
    at Object.InvalidProvider (web3.min.js:2:962448)


```

**Comment by glitch-txs** [comment]
Your code looks fine, which wallet are you using?

**Comment by mostafa-rio** [comment]
I use MetaMask mobile app to connect. but thing is MetaMask and other wallets like trust wallet is not in the pop up of the web3modal. and think that might be also an issue 🤔

**Reply by glitch-txs** [reply]
iOS?

Could you test here?
https://lab-walletconnect-modal.pages.dev/with-ethereum-provider

**Reply by mostafa-rio** [reply]
I tested it and works great. Is this made using react? Could share the code, please?
(I'm using Android. but there is no difference I guess ).

**Reply by glitch-txs** [reply]
it's from labs:
https://github.com/WalletConnect/modal/blob/main/laboratory/src/pages/with-ethereum-provider/index.tsx

you can also check the examples repo:
https://github.com/WalletConnect/web3modal-examples

**Reply by mostafa-rio** [reply]
Checked the examples. I can't see any examples with web3js and also no example in Angular of course.

**Reply by glitch-txs** [reply, marked as answer]
Just tested it, your code is correct, it seems like MetaMask doesn't support adding more than one chain. You can try this:

````ts
const provider = await EthereumProvider.init({
      projected: 'projectid',
      chains: [1],
      showQrModal: true,
      optionalChains: [137, 56]
      //other config...
})

**Reply by mostafa-rio** [reply]
Yes, it works. Man, you're a pro in web3 :). I really appreciate the time you put in.
(Still has an issue with signing messages using only with MetaMask on mobile)

**Reply by glitch-txs** [reply]
Thanks! I'd open an issue in MetaMask mobile repo, this is also an issue I think, they should support multiple chains 🤔

**Reply by mostafa-rio** [reply]
check this out: 
https://github.com/MetaMask/metamask-mobile/issues/6688

**Reply by TEE2DWHY** [reply]
Hello @mostafa-rio . Thanks for this. Pls does the sendTransaction method work in trust wallet ??. I’m presently having an issue with sending transaction using wallet connect etheruem provider. It works in safepal and metamask but not trust wallet. I’ll appreciate your response. Thank you

**Reply by nexus6nexus6** [reply]
This solution poses other issues down the road.

If you use `optionalChains` to get your MM wallet connected you'll run into issues with switching chains.
I'm getting `"Error: Failed to switch to chain 'eip155:1'. The chain is not approved or the wallet does not support 'wallet_switchEthereumChain' method"`  or you'll just get a positive resolution (`null`) but the chain in the Metamask won't change at all.

Metamask thread: https://github.com/MetaMask/metamask-mobile/issues/6688

**Comment by nexus6nexus6** [comment]
> Just tested it, your code is correct, it seems like MetaMask doesn't support adding more than one chain. You can try this:
> 
> ```ts
> const provider = await EthereumProvider.init({
>       projected: 'projectid',
>       chains: [1],
>       showQrModal: true,
>       optionalChains: [137, 56]
>       //other config...
> })
> ```

Be aware.  Using `optionalChains` will get you in trouble down the line.  Switch chains method won't work on MM and on other wallets connected using this approach. Might be other methods won't work as well. See here: [link](https://github.com/MetaMask/metamask-mobile/issues/6688#issuecomment-1647846096)

**Reply by glitch-txs** [reply]
Hi, adding optionalChains should be supported by all wallets, have you find any issue with other wallets?
